### PR TITLE
feat(swingset): get queues length

### LINKED ISSUE #5238: add controller.getRunQueueSize()
## What is the Problem Being Solved?

To support #5176, in which cosmos will stop accepting messages into the mempool and into blocks once the kernel announces it is too full, we need to give the kernel a voice. @michaelfig says it'd be sufficient to let cosmic-swingset poll the kernel for it's queue depth.

## Description of the Design

Add `controller.getRunQueueSize()` (feel free to bikeshed the name). Returns a set of Nat (non-negative integer), one for active queues, one for inactive queues (such as promise queues). Only safe/consistent to call when the kernel is idle (i.e. outside of a `controller.run()`).

Caller is responsible for deciding how to interpret the result. 0 means empty, but the "full" threshold is up to the caller (and should be configurable). Once priority queues are introduced (#4318), this would need to be updated to expose different numbers for low and high priority queues, and the caller may pay attention to two thresholds.

## Security Considerations

None I can think of.

## Consensus Considerations

The kernel-reported value will grow in response to kernel device IO calls (e.g. adding messages to the Mailbox device will push work onto the kernel queue, increasing the number). Calling `controller.run()`, with some particular `runPolicy`, on a non-empty queue could either increase or decrease the number (but in the long run should decrease, assuming no runaway computation).

So this should be safe to use for consensus operations as long as all replicas use the same sequence of device writes and `controller.run()` calls, and use the same `runPolicy`, and refrain from calling it concurrently with a `controller.run()` (which might sample the queue at random points during execution, which would give inconsistent results).

~~However we don't currently plan to use this for consensus operations: accepting messages into the mempool is done separately by each validator and is outside of consensus.~~

We plan to use this mechanism for both consensus (validating a proposed block) and non consensus (accepting into the mempool)

## Test Plan

Unit tests, probably in `test-kernel.js`.

### PR DESCRIPTION
**Pull request by mhofman**
closes: #5238
refs: #5176

## Description

This PR tracks the number of messages pending on promise queues as a new kernel stat, and exposes the run-queue + acceptance queue length, as well as the promise queue length in a new controller method, which can be used by cosmic-swingset to decide if messages should be delivered to mailbox or possibly bridge devices.

To make sure the stats are consistent with the rest of the kernel state, this PR also:
- Split the kernel stats into consensus and local sections, saved separately (partially undoes https://github.com/Agoric/agoric-sdk/pull/3625/commits/df8359eca80e28736d294a558ed6c5e3b8b14127)
- Commits / rolls-back the `kernelStats` along kernel state, but avoiding intermediary DB writes

### Security Considerations

The `kernelStats` were previously out of consensus and not thoroughly vetted to be deterministic, or consistent with the kernel state. I believe I have audited all changes to the stats, especially the ones that are non in consensus, and made sure any crank abort or reload would reflect in the stats kept in RAM.

### Documentation Considerations

Description added, stats don't seem to be documented anywhere else than in `metrics.js`.

I have now fully typed the metrics allowing calls like `incStats` to type check that the stat name is valid.

### Testing Considerations

I added unit tests for the stats logic, however nothing tests the content of stats against expectations regarding the rest of the kernel state.

One example are the up/down stats. The base value should always be a pure function of the known state of the kernel, but nothing enforces or verifies that. I've manually reviewed all paths where existing run-queue and acceptance queue, as well as the new promise queue, are modified to ensure stats are tracked appropriately

Putting some of these stats under consensus should quickly highlight non deterministic stats (if any) by the deployment integration test. Swingstore traces in integration should also highlight divergences in non-consensus stats however such consistency is not yet enforced in the integration test due to snapshot hash divergences.

### CODE REVIEW

**Comment by warner** [comment]
API seems good.. having it depend upon `stats` feels sligtly weird, but maybe that's just because of the not-intended-for-consensus property that you pointed out. OTOH it'd be a shame to duplicate code to get this data.

I'm not 100% sure the stats are managed accurately across crank unwinding and/or replay, I seem to remember them being held in RAM for some period of time, and then getting written to the DB only at the end of a crank, and I remember being nervous about that but decided that since they weren't being used for consensus, it wouldn't matter.

So maybe it *would* be a good idea to track them separately. The run-queue and acceptance-queue are still single keys, we could just JSON.parse and `.length` them when requested (which won't be very often, assuming cosmic-swingset only asks once per block, but OT3H maybe that's a bad assumption to make).

Dunno.

**Comment by mhofman** [comment]
I did double check the write logic, and it seemed sane. I will double check the abort crank logic however, as now I'm not sure if we reload the saved object at that point.

That said, I think it should be safe to use for our purpose:
- Accepting messages into the mem-pool is non-consensus in the first place
- Accepting messages into a block is done by the block proposer, which I think means it's also out of consensus in the sense other nodes would accept the proposer's decision?

I'm happy to move a subset of stats explicitly into consensus. Not sure the best approach for that is. I already have reservations about all the transient serializations and deserializations happening during a crank, and moving stats under that mechanism doesn't feel quite right.

**Comment by mhofman** [comment]
I did find issues related to crank abort, and when stats are loaded or first initialized. I have now aligned them with other store entries, and moved them in consensus, including participating in the crankhash. They are however still only written down at the end of the crank to avoid DB churn.

File: packages/SwingSet/src/kernel/kernel.js [hunk 1]
```diff
@@ -1645,6 +1643,27 @@ export default function buildKernel(
       return kernelKeeper.getActivityhash();
     },
 
+    // While stats are local and technically not part of concensus state, these
```
Comment (warner) [inline packages/SwingSet/src/kernel/kernel.js]: typo s/concensus/consensus/

also let's say ".. and not enforced as part of consensus state, we would expect all replicas to have the same value", to make it clear that this equality is an expectation, not a demand

File: packages/SwingSet/test/test-state.js [hunk 2]
```diff
@@ -701,6 +703,22 @@ test('meters', async t => {
   t.deepEqual(k.getMeter(m3), { remaining: 'unlimited', threshold: 5n });
 });
 
+const makeTestCrankHasher = k => {
```
Comment (warner) [inline packages/SwingSet/test/test-state.js]: This is ok, but a big purpose of this test is to ensure we really understand what's going into the activityhash, which is why it was doing `h.update()` with literal or quasiliteral strings: minimal indirection, maximum visibility of the hash inputs.

Having kernelStats be part of the consensus state makes that approach pretty annoying, so I completely get why this layer of indirection is useful, and it's fine, just keep in mind the reason for the original structure.

**Comment by warner** [comment]
Yeah, I'm dubious about enforcing consensus of kernelStats. It would prevent us from adding new stats in minor kernel updates (those updates would have to be deployed at exactly the same time across all replicas), as well as stats on things that really are replica-local (like the number of times the kernel has been rebooted, or heap snapshot activity, or some sort of cache instrumentation).

Until now (where we want to use some stats for backpressure), these stats have been entirely for telemetry/debugging purposes. Our current plan for backpressure is to use it in a non-consensus fashion (mempool inlet control).

OTOH (and @michaelfig can probably clarify), it might be easy to accidentally (or intentionally) implement the cosmic-swingset -side backpressure queue-depth check in a way that *is* part of consensus (at the beginning of DeliverTx, causing a txn error if the block proposer allows a swingset message in despite the queues being too full).

Part of me wants two separate kernelStats, one local (with most of the stats), and the other in consensus (only containing the queue depths). Then we don't need to worry about cosmic-swingset accidentally relying upon something that swingset doesn't guarantee consensus about. But having two sets of stats feels like bloat, especially if they both need `initializeStats` and updates after each crank.

Maybe there's a deeper problem with thinking of SwingStore as an acceptable place to hold non-consensus data at all, and we should provide the kernel with a secondary non-consensus storage tool instead of using `local.*` to distinguish them. If we ever wanted to try (again) to store swingset data in the cosmos-sdk IAVL tree, the `local.` trick wouldn't work, and we'd definitely need a secondary DB. But that sounds like far more work than we can do for MN-1.

Hrmph. So the short-term options are:
* 1: include kernelStats in consensus
  * risks a MN-1 consensus failure if one of our stats is less stable than we thought (I'm looking suspiciously at `kernelDevices`, probably ok but hmm)
  * adding new non-consensus stats would require more work in the future (but maybe YAGNI)
* 2: split into `kernelStats` and `local.kernelStats`
  * somewhat less risk
  * more work now, kinda annoying bloat

@FUDCo any opinions here?

**Comment by mhofman** [comment]
> Part of me wants two separate kernelStats, one local (with most of the stats), and the other in consensus (only containing the queue depths). Then we don't need to worry about cosmic-swingset accidentally relying upon something that swingset doesn't guarantee consensus about. But having two sets of stats feels like bloat, especially if they both need `initializeStats` and updates after each crank.

I think 2 stats is right when we need them, and when I audited the current stats, they all seemed to be based on consensus state already, so I didn't see the harm in moving them all into consensus. I'm not sure 2 different keys, one prefixed by `local.` is that much bloat.

> * I'm looking suspiciously at `kernelDevices`

If we believe this may be different between validators, I'm happy to review, but the devices are already stored under the `device.` and `d$NN` keys in the store, so they're already part of consensus.

**Comment by FUDCo** [comment]
> @FUDCo any opinions here?

I'd also go with option 2.  I think it's proper to distinguish between stats that are kept for performance monitoring, tuning, and debugging vs. stats that are part of the operational logic of the chain.

**Comment by mhofman** [comment]
@warner @FUDCo PTAL

File: packages/SwingSet/src/controller/controller.js [hunk 3]
```diff
@@ -368,7 +368,7 @@ export async function makeSwingsetController(
         `alleged bundleID ${allegedBundleID} does not match actual ${bundleID}`,
       );
     }
-    kernel.installBundle(bundleID, bundle);
+    await kernel.installBundle(bundleID, bundle);
```
Comment (warner) [inline packages/SwingSet/src/controller/controller.js:371]: good catch, I think the old version was sync

File: packages/SwingSet/test/test-state.js [hunk 4]
```diff
@@ -701,6 +707,27 @@ test('meters', async t => {
...
+const { consensusStats: testCrankHasherConsensusStats } =
+  testCrankHasherStats.getSerializedStats();
+
+const makeTestCrankHasher = () => {
+  const h = createHash('sha256');
+  const update = (...args) => {
+    h.update(args.map(val => `${val}\n`).join(''));
+  };
+  const digest = () => {
+    update('add', 'kernelStats', testCrankHasherConsensusStats);
```
Comment (warner) [inline packages/SwingSet/test/test-state.js]: yeah ok, this is a decent compromise, but maybe add a note to remind ourselves that this is hard-wired to include only the initial value of the kernel stats, so any new test that calls a `kernelKeeper` function that increments a state will get the wrong crankhash

**Review by warner — APPROVED** [review]
Looks good, minor comment suggestion

File: packages/SwingSet/src/kernel/kernel.js (hunk 1 above)
Comment (mhofman) [reply packages/SwingSet/src/kernel/kernel.js]: That comment has been removed already.

File: packages/SwingSet/test/test-state.js (hunk 2 above)
Comment (mhofman) [reply packages/SwingSet/test/test-state.js]: I'd be happy to revert the string serialization part.

**Comment by mhofman** [comment]
@warner, I addressed your feedback in e16a9a3972f419ad7b9a43a10d278f6b4b60aeb4 then squashed the PR in a way that hid all the back and forth in this PR for a clean merge, while preserving individual commits that make sense: [`e16a9a3..e10564d`](https://github.com/Agoric/agoric-sdk/compare/e16a9a3972f419ad7b9a43a10d278f6b4b60aeb4..e10564d2e06a571e6d4f2bfd9a4c1ae9f8e622ee)

File: packages/SwingSet/test/test-state.js (hunk 2 above)
Comment (mhofman) [reply packages/SwingSet/test/test-state.js]: Reverted

**Pull request merged**